libcaf_core/CMakeLists.txt

```diff
@@ -131,33 +127,64 @@ set(LIBCAF_CORE_SRCS
   src/work_stealing.cpp
 )
 
+# configure build_config.hpp header
+macro(to_int_value name)
+  if(${name})
+    set(${name}_INT 1)
+  else()
+    set(${name}_INT -1)
+  endif()
+endmacro()
+
+to_int_value(CAF_NO_EXCEPTIONS)
+to_int_value(CAF_NO_MEM_MANAGEMENT)
+to_int_value(CAF_ENABLE_RUNTIME_CHECKS)
+
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/../cmake/build_config.hpp.in"
+               "${CMAKE_CURRENT_BINARY_DIR}/caf/detail/build_config.hpp"
+               @ONLY)
+install(FILES "${CMAKE_CURRENT_BINARY_DIR}/caf/detail/build_config.hpp"
+        DESTINATION include/caf/detail
+)
+
 add_custom_target(libcaf_core)
 
 # build shared library if not compiling static only
 if (NOT CAF_BUILD_STATIC_ONLY)
   add_library(libcaf_core_shared SHARED ${LIBCAF_CORE_SRCS} ${LIBCAF_CORE_HDRS})
   target_link_libraries(libcaf_core_shared ${CAF_EXTRA_LDFLAGS})
+  target_include_directories(libcaf_core_shared PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:include>
+  )
   set_target_properties(libcaf_core_shared
                         PROPERTIES
                         SOVERSION ${CAF_VERSION}
                         VERSION ${CAF_VERSION}
                         OUTPUT_NAME caf_core)
-  if (CYGWIN)
-    install(TARGETS libcaf_core_shared RUNTIME DESTINATION bin)
-  else()
-    install(TARGETS libcaf_core_shared LIBRARY DESTINATION lib)
-  endif()
+  install(TARGETS libcaf_core_shared
+    RUNTIME DESTINATION bin
+    LIBRARY DESTINATION lib
+  )
   add_dependencies(libcaf_core_shared libcaf_core)
 endif ()
 
 # build static library only if --build-static or --build-static-only was set
 if (CAF_BUILD_STATIC_ONLY OR CAF_BUILD_STATIC)
   add_library(libcaf_core_static STATIC ${LIBCAF_CORE_HDRS} ${LIBCAF_CORE_SRCS})
   target_link_libraries(libcaf_core_static ${CAF_EXTRA_LDFLAGS})
+  target_include_directories(libcaf_core_static PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:include>
+  )
   set_target_properties(libcaf_core_static PROPERTIES OUTPUT_NAME caf_core_static)
   install(TARGETS libcaf_core_static ARCHIVE DESTINATION lib)
   add_dependencies(libcaf_core_static libcaf_core)
 endif ()
 
-link_directories(${LD_DIRS})
-include_directories(. ${INCLUDE_DIRS})
+install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/caf"
+        DESTINATION include
```

Comment:
Aside from the discussion above (either move it back or consistently place the installs in the submodule CMake files), this now installs all CAF headers into the wrong destination. Should read `DESTINATION include/caf`.

Reply:
Note the removed directory separator at the end of the `DIRECTORY` argument, this way the source directory itself gets copied into the destination. The old form of the call used `libcaf_core/caf/` as the argument, which is handled by copying every member inside that source into `DESTINATION`.

Reply:
Gotya, thanks. Sometimes it isn't easy to think of all the subtleties in a review.